docs/design_spec.md

```diff
@@ -0,0 +1,94 @@
+# Nimbess Agent Design
+
+## Overview
+
+Nimbess Agent's main purpose is to manage an external data plane. The Agent's main responsibilities
+include:
+
+ 1. Translating Northbound (NB) application-specific resources into network oriented resources across the
+ entire network fabric.
+ 2. Managing and serving as the control plane for the locally assigned, and external data plane (Southbound).
+ 
+ The Agent is distributed over every host and each Agent instance is capable of viewing and managing the
+ network configuration for the entire fabric. However, only the Agent local to a data plane (on the same host)
+ is capable of directly configuring that data plane.
+ 
+## Agent Northbound Design
+ 
+Nimbess Agent is able to support multiple container platforms and by default supports Kubernetes (K8S). It
+also supports the CNI spec for handling CNI requests from Kubelet or other container platforms which use
+CNI. The Nimbess Agent NB is responsible for translating K8S resources into network-oriented resources.
+
+The Agent exposes a gRPC based CNI API in order to service CNI requests. The requests are translated into
+Nimbess Ports, and a Network Pipeline is built for that Port. For other networking related container platform
+resources, such as a K8S Service, Ingress, etc; Nimbess Agent will handle translating those resources into their
+relevant network-oriented constructs and applied to one or more pipelines.
+ 
+## Network Constructs in Nimbess Agent
+ 
+In Nimbess, how traffic flows through a data plane can be though of as a pipeline of Network Functions.
+Traffic ingresses a data plane, flows through a series of network functions such as a Switch, ACL,
+Load-Balancer, etc and then egresses the data plane out of a port. Nimbess uses this concept of a Network
+Pipeline and creates one for each Port that is attached to the data plane. There are multiple reasons for
+this design:
+ 
+ 1. Guarantees loop-free forwarding
+ 2. Increases per network function processing efficiency
+ 3. Simplifies debugging
+ 
+When a Port is added to a pipeline, it is broken into an IngressPort as well as an EgressPort. Only the
+IngressPort is ever attached to a Port's pipeline. This prevents traffic from ever being forwarded back
+out the same interface from which it originated (1). The EgressPort is attached to every other port
+pipeline which belongs to the same network fabric (L2/L3 domain). Additionally, this design increases network
+function processing efficiency, because the cache of the tables are limited to lookups for traffic that may ingress
+a single port. Debugging becomes easier because the only traffic in a particular pipeline only ever comes
+from a single port. However, one of the drawbacks to this design is that there is increased memory usage,
```

Comment:
Do you have a feel for the amount of memory consumed for a typical pipeline?
I'm just wondering how this design will scale. As you mention in the spec, there is a pipeline per port. So, memory usage is something like (mem per pipeline x number of ports) and the pipeline mem usage is dependent on what services/net functions are configured.
I can see folks asking about this early on, more of a heads up that I think it would be good to have a feel for this early on, in case we need to think about other options with regards to scaling.

Reply:
I think we should be OK memory wise with nimbess-agent. I think in BESS the question will be how does it scale with this many pipelines. It is something we still need to investigate with some scale testing after we implement this agent. Right now we do not have an easy way via a control plane to do that scaling, but with this agent we will have the means.